README.md
# Morse-Converter-Windows-App-Cpp
Semester project of Programming 2 (C++)


history.txt
tushig
 - ..- ... .... .. --.

_________________________________________________


Morse.h
#ifndef MORSE_H
#define MORSE_H
//#pragma once

#include <iostream>
#include <string.h>

//Initializing binary tree using linked list
typedef struct BinaryTree {
	char characters;
	int key;
	struct BinaryTree* left;
	struct BinaryTree* right;
}
BinaryTree;

//Initializing the structure of library contains characters and strings
typedef struct MorseLibrary {
	char letter;
	const char* MorseCode;
}
MorseLibrary;




//Iinitializing essential class of this project
//It does main conversions
class Morse {
private:
	BinaryTree* start = NULL;
	//Initializing array with all characters and their translations in Morse code
	MorseLibrary characters[49] = {
		{'a', ".-"},
		{'b', "-..."},
		{'c', "-.-."},
		{'d', "-.."},
		{'e', "."},
		{'f', "..-."},
		{'g', "--."},
		{'h', "...."},
		{'i', ".."},
		{'j', ".---"},
		{'k', "-.-"},
		{'l', ".-.."},
		{'m', "--"},
		{'n', "-."},
		{'o', "---"},
		{'p', ".--."},
		{'q', "--.-"},
		{'r', ".-."},
		{'s', "..."},
		{'t', "-"},
		{'u', "..-"},
		{'v', "...-"},
		{'w', ".--"},
		{'x', "-..-"},
		{'y', "-.--"},
		{'z', "--.."},
		{'0', "-----"},
		{'1', ".----"},
		{'2', "..---"},
		{'3', "...--"},
		{'4', "....-"},
		{'5', "....."},
		{'6', "-...."},
		{'7', "--..."},
		{'8', "---.."},
		{'9', "----."},
		{':', "--..."},
		{',', "--..--"},
		{'(', "-.--."},
		{')', "-.--.-"},
		{'!', "-.-.--"},
		{'/', "-..-."},
		{'=', "-...-"},
		{'-', "-....-"},
		{'@', ".--.-."},
		{'+', ".-.-."},
		{'.', ".-.-.-"},
		{'"', ".-..-."},
	};
public:
	Morse();
	~Morse();
	BinaryTree* Node(char item, int key);
	BinaryTree* AddNode(BinaryTree* start, char value, int key);
	char GetText(const char* str);
	const std::string MorseToText(const char* str1);
	const std::string TextToMorse(const char* txt);
	void statistics();
	void deleteNode(BinaryTree* root);
};

#endif





Audio.cpp
#include "Audio.h"

//This function is initializing a sound for dot
void Audio::dotAudio()
{
	Beep(900, 300);
}
//This function is initializing a sound for dash
void Audio::dashAudio()
{
	Beep(900, 500);
}
//This function is for pausing or seperating the sound between dot and dash
void Audio::pauseAudio()
{
	Sleep(500);
}
//Main function that is making it possible to make sound from the string by finding dot and dash
void Audio::MorseBeeper(std::string str)
{
	int length = str.length();
	for (int i = 0; i < length; i++)
	{
		if (str[i] == '.')					//if it finds dot it will play dotAudio()
		{
			dotAudio();
		}
		else if (str[i] == '-')				//if it finds dot it will play dashAudio()
		{
			dashAudio();
		}
		else
		{
			pauseAudio();					//if it didn't find - or . it will pause the sound
		}
	}
}
/**
	This function is initializing listen text file to store morse code
	because of windows application cannot get the previously converted
	morse code by passing it within the cases. The comment will be
	detailed in the main.cpp
*/
void Audio::listen(char* insertMorse)
{
	listenAudio.open("listen.txt");
	listenAudio << insertMorse;
	listenAudio.close();

}

Morse.cpp
#include "Morse.h"
//Node function is making it possible to set aside the characters of the binary tree
BinaryTree* Morse::Node(char item, int key)
{
	BinaryTree* tmp = new BinaryTree[sizeof(BinaryTree)];

	tmp->characters = item;
	tmp->key = key;
	tmp->left = NULL;
	tmp->right = NULL;
	return tmp;
}
//This function is for adding nodes
//It finds characters by if its key value goes right or left 
BinaryTree* Morse::AddNode(BinaryTree* start, char value, int key)
{
	if (start == NULL)
		return Node(value, key);
	if (key < start->key)
		start->left = AddNode(start->left, value, key);
	else if (key > start->key)
		start->right = AddNode(start->right, value, key);
	return start;
}
//Constructor has binary tree
Morse::Morse()
{
	char characters[61] = " tmo#09#8:gqz#,7nky()c#!dx/b=6-eawj1'p#@r#+.l#&iu#2#?fsv3h45";	//Available characters in order of their insertion
	int keys[61] = { 31, 15, 8, 4, 2, 1, 3, 5, 6, 7, 10, 9, 13, 12, 11, 14, 23, 19, 16, 18, 17, 22, 21, 20, 26, 24, 25, 28, 27, 30, 29, 47, 39, 35, 34, 32, 33, 38, 36, 37, 43, 40, 42, 41, 45, 44, 46, 54, 52, 49, 48, 50, 51, 53, 57, 56, 55, 59, 58, 60 }; /*key values of characters*/
	start = AddNode(start, characters[0], keys[0]);
	for (int i = 1; i < 61; i++)															//inserting all available characters to tree using AddNode
	{
		AddNode(start, characters[i], keys[i]);
	}
}
//Deconstructor
Morse::~Morse()
{
	deleteNode(start);
}
char Morse::GetText(const char* str)
{
	int size = strlen(str);						//getting length of the str
	BinaryTree* pointer = start;
	for (int i = 0; i < size; i++)
	{
		if (str[i] == ' ')						//| is space
		{
			break;
		}

		if (str[i] == '-')						//dash is left
		{
			pointer = pointer->left;
		}

		if (str[i] == '.')						//dot is right
		{
			pointer = pointer->right;
		}

		if (pointer == NULL)					//if start couldn't find anything it will return 0
		{
			printf("Invalid code!\n");
			return 0;
		}
	}
	return pointer->characters;
}

//Essential function to convert morse code input to text
const std::string Morse::MorseToText(const char* str1)
{
	char* str_token, * token1;
	std::string output = "";
	char* tmp = new char[strlen(str1)];
	for (int i = 0; i < strlen(str1); i++)
	{
		tmp[i] = str1[i];
	}
	str_token = strtok_s(tmp, " ", &token1);		//I had to make it return to a pointer to the first character
													//using strtok is splitting the string 
	while (str_token != NULL)
	{
		output = output + GetText(str_token);
		str_token = strtok_s(NULL, " ", &token1);	//getting spaces between morse codes
	}
	return output;									//Initializing it by non-void function will be easier to display its output in windows application
}

void Morse::statistics()
{

	for (int i = 0; i < 49; i++)					//print all statistics
	{
		printf("%c < -> %s\n", characters[i].letter, characters[i].MorseCode);
	}
}

//Deleting the node function
void Morse::deleteNode(BinaryTree* root)
{
	if (root == NULL)								//if already empty return
	{
		return;
	}

	delete[] root->left;
	delete[] root->right;
	delete[] root;
}


//Essential function to convert text input to Morse code
const std::string Morse::TextToMorse(const char* txt)
{
	int length = strlen(txt);
	std::string output = "";
	int StringLen = 0;
	for (int i = 0; i < length; i++)
	{
		for (int j = 0; j < 49; j++)									//comparing the characters and morse code
		{
			if (txt[i] == characters[j].letter)
			{
				StringLen = +strlen(characters[j].MorseCode);

				const char* morseStr = new char[StringLen];
				morseStr = characters[j].MorseCode;
				output = output + " " + characters[j].MorseCode;		//it will return output with " " space

			}
		}
	}

	return output;														//The main reason of returning it will make it easier to display in the windows application
}

main.cpp
/**
	I am using <windows.h> library for making it windows application.
	<tchar.h> is for defining the char which resolves ultimately
	to wchar_t. If it is built with UNICODE enabled we do not need to
	use tchar.
*/
#define _CRT_SECURE_NO_WARNINGS		// Remove secure warnings 

#include "Morse.h"
#include <windows.h>
#include "Audio.h"
#include "History.h"
#include <tchar.h>
#include <stdio.h>

//My program has 4 buttons and I defined them:
#define LISTEN_BUTTON 2
#define HISTORY_BUTTON 3
#define CONVERT_BUTTON 4
#define NEWCONVERT_BUTTON 5

const char g_szClassName[] = "Tushig's Morse Converter";				//Title of the program

LRESULT CALLBACK WndProc(HWND, UINT, WPARAM, LPARAM);
void AddButtons(HWND);
HWND hEdit, hText, hMorse, hOutput, textFile;							//Declarations of the textbox windows

int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpszArgument, int nCmdShow)
{
	HWND hwnd;			//Handle for the window
	MSG messages;		//Messages to the saved application
	WNDCLASSEX winclass;	//Data struct for the windows class

	//Window stucture:
	winclass.hInstance = hInstance;
	winclass.lpszClassName = g_szClassName;
	winclass.lpfnWndProc = WndProc;			//function called by windows
	winclass.style = CS_DBLCLKS;				//catch double clicks
	winclass.cbSize = sizeof(WNDCLASSEX);

	//Default icon and mouse pointers:
	winclass.hIcon = LoadIcon(NULL, IDI_APPLICATION);
	winclass.hIconSm = LoadIcon(NULL, IDI_APPLICATION);
	winclass.hCursor = LoadCursor(NULL, IDC_ARROW);
	winclass.lpszMenuName = NULL;				//no menu
	winclass.cbClsExtra = 0;					//no extra bytes after the window class
	winclass.cbWndExtra = 0;					//windows instance structure

	//Using windows's default color as the background color of the window
	winclass.hbrBackground = (HBRUSH)GetStockObject(LTGRAY_BRUSH);	//COLOR_BACKGROUND;

	//Registering the window class, and it quits if fail
	if (!RegisterClassEx(&winclass))
		return 0;

	//Creating the application

	hwnd = CreateWindowEx(
		0,									//extendend possibilities for variation
		g_szClassName,						//class name
		"Tushig's Morse Converter",			//Title of the top
		WS_SYSMENU | WS_MINIMIZEBOX,		//default window
		CW_USEDEFAULT,						//windows's position
		CW_USEDEFAULT,						//where windows ends up on display
		800,								//width
		500,								//height
		HWND_DESKTOP,						//app is child window to the display
		NULL,								//no menu
		hInstance,							//Instance handler
		NULL);								//No window creatiom data

	//Making the window visible on display
	ShowWindow(hwnd, nCmdShow);

	//run the message loop until GetMessage() returns 0
	while (GetMessage(&messages, NULL, 0, 0))
	{
		while (GetMessage(&messages, NULL, 0, 0))
		{
			TranslateMessage(&messages);		//translate virtual key to character
			DispatchMessage(&messages);			//send message to WndProc
		}
		return messages.wParam;
	}

}


//It is called by the DispatchMessage() function
LRESULT CALLBACK WndProc(HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam)
{
	switch (message)			//handling message
	{
	case WM_CREATE:				//Window will be initialized in this case

		AddButtons(hwnd);		//AddButtons function has my program's buttons and textboxes (I initialized it in the end of this file)
		break;
	case WM_COMMAND:			//Operations and functionalities of buttons will be initialized in this case
		switch (LOWORD(wParam))
		{
		case CONVERT_BUTTON: //It is for converting text to morse
		{
			Morse Morse1;
			History history;
			Audio aud;

			history.createHistory();
			char text[50], outText[50];

			GetWindowText(hText, text, 50);			//Getting user input from hText textbox
			try
			{
				if (strlen(text) == 0)
				{
					throw 0;
				}
				else
				{
					history.addToHistory(text);
					strcpy_s(outText, Morse1.TextToMorse(text).c_str());		//Converting text to morse and copying it into the outText

					SetWindowText(hOutput, outText);							//Displaying outText in the hOutput textbox
					history.addToHistory(outText);								//adding to history
					history.lineBreak();
					history.closeFile();
					aud.listen(outText);										//Storing converted morse to the listen.txt file
				}
			}
			catch (int num)
			{
				std::string tmp = "Invalid! The text box is empty!";
				SetWindowText(hOutput, tmp.c_str());
			}
			break;
		}

		case NEWCONVERT_BUTTON:						//It is for converting morse into the text it has same functionality with the previous case 
		{

			Morse Morse2;
			History history;

			char  morse[50], outMorse[50];

			GetWindowText(hMorse, morse, 50);
			try
			{
				if (strlen(morse) == 0)
				{
					throw 0;

				}
				else
				{
					history.addToHistory(morse);

					strcpy_s(outMorse, Morse2.MorseToText(morse).c_str());
					SetWindowText(hOutput, outMorse);
					history.addToHistory(outMorse);
					history.lineBreak();

					history.closeFile();
				}
			}
			catch (int num)
			{
				std::string tmp = "Invalid! The text box is empty!";
				SetWindowText(hOutput, tmp.c_str());

			}
			break;
		}
		case HISTORY_BUTTON:									//Show message box if the user tap history button
		{
			::MessageBox(hwnd, "History saved automatically in the file directory", "Alert!", MB_OK);
			break;
		}
		case LISTEN_BUTTON:										//It is for listening the converted morse code
		{
			std::string audioL;
			std::ifstream file;
			Audio sound;
			file.open("listen.txt");							//opening and reading from listen.txt file
			getline(file, audioL);
			sound.MorseBeeper(audioL);							//MorseBeeper function will play the sound by finding dot and dashes

			break;
		}
		}
		break;

	case WM_DESTROY:
		PostQuitMessage(0);          // Exit from the program
		break;
	default:
		return DefWindowProc(hwnd, message, wParam, lParam);
	}
	return 0;
}

//I initialized buttons and textbox windows in this function
void AddButtons(HWND hWnd)
{

	CreateWindow("Static", "Enter Text Here:", WS_VISIBLE | WS_CHILD, 100, 50, 100, 40, hWnd, (HMENU)1, NULL, NULL);
	hText = CreateWindow("Edit", "", WS_VISIBLE | WS_CHILD | WS_BORDER, 200, 50, 180, 40, hWnd, (HMENU)1, NULL, NULL);

	CreateWindow("Static", "Enter Morse Code Here:", WS_VISIBLE | WS_CHILD, 100, 90, 100, 40, hWnd, (HMENU)1, NULL, NULL);
	hMorse = CreateWindow("Edit", "", WS_VISIBLE | WS_CHILD | WS_BORDER, 200, 90, 180, 40, hWnd, (HMENU)1, NULL, NULL);

	CreateWindow("Button", "Convert to Morse", WS_VISIBLE | WS_CHILD, 100, 140, 120, 40, hWnd, (HMENU)CONVERT_BUTTON, NULL, NULL);

	CreateWindow("Button", "Convert to Text", WS_VISIBLE | WS_CHILD, 260, 140, 120, 40, hWnd, (HMENU)NEWCONVERT_BUTTON, NULL, NULL);

	hOutput = CreateWindow("Edit", "Result:", WS_VISIBLE | WS_CHILD | WS_BORDER, 450, 50, 250, 80, hWnd, (HMENU)1, NULL, NULL);

	textFile = CreateWindow("STATIC", "\nStatistics:"
		"\na .-, b -..., c -.-., d -.., e ., f ..-., g --., h ...., i .., j .---, k -.-, l .-.., m --, n -., o ---, p .--."
		"q --.-, r .-., s ...,  t -, u ..-, v ...-, w .--, x -..-, y -.--, z --..\n"
		"\nNumbers:"
		"\n0 - - - - -, 1 .- - - -, 2 ..- - -, 3 ...- -, 4 ....-, 5 ....., 6 -...., 7 - -..., 8 - - -.., 9 - - - -.\n"
		"\nCharacters:"
		"\n: --..., , --..--, ( -.--., ) -.--.-, ! -.-.--, / -..-., = -...-, - -....-, @ .--.-., + .-.-., . .-.-.-, '' .-..-.\n", WS_VISIBLE | WS_CHILD | WS_BORDER, 100, 200, 600, 200, hWnd, NULL, NULL, NULL);

	CreateWindow("Button", "History", WS_VISIBLE | WS_CHILD, 450, 140, 110, 40, hWnd, (HMENU)HISTORY_BUTTON, NULL, NULL);

	CreateWindow("Button", "Listen", WS_VISIBLE | WS_CHILD, 590, 140, 110, 40, hWnd, (HMENU)LISTEN_BUTTON, NULL, NULL);

}


//References:
/**
	http://eik.bme.hu/~csurgai/Mswin_en/Mswin_en.htm
	https://www.cplusplus.com/reference/string/string/c_str/
	https://www.dropbox.com/s/k7fb15c3f1o4yf8/forgers-win32-tutorial.pdf?dl=0
	https://docs.microsoft.com/en-us/windows/win32/learnwin32/winmain--the-application-entry-point
	http://www.winprog.org/tutorial/controls.html
	https://www.youtube.com/watch?v=FXrulJXMiU8
	https://www.youtube.com/c/ThePentamollisProject/videos
*/